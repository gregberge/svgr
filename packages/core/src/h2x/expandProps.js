import { JSXAttribute } from 'h2x-plugin-jsx'

const expandProps = () => () => ({
  visitor: {
    JSXElement: {
      enter(path) {
        if (
          path.node.name === 'svg' &&
          !path.node.attributes.some(attr => attr && attr.name === 'props')
        ) {
          const props = new JSXAttribute()
          props.name = 'props'
          props.spread = true
          path.node.attributes.push(props)
          path.replace(path.node)
        }
      },
    },
  },
})

export default expandProps
