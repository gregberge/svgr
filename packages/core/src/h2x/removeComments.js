const removeComments = () => () => ({
  visitor: {
    JSXComment: {
      enter(path) {
        path.remove()
      },
    },
  },
})

export default removeComments
