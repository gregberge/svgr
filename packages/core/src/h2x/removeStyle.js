const removeStyle = () => () => ({
  visitor: {
    JSXElement: {
      enter(path) {
        if (path.node.name === 'style') {
          path.remove()
        }
      },
    },
  },
})

export default removeStyle
