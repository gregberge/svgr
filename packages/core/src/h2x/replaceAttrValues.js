const replaceAttrValues = (attributes = {}) => () => ({
  visitor: {
    JSXAttribute: {
      enter(path) {
        Object.keys(attributes).forEach(key => {
          if (path.node.value === key) {
            path.node.value = attributes[key]
          }
        })
      },
    },
  },
})

export default replaceAttrValues
