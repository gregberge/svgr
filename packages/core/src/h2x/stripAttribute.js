const stripAttribute = attribute => () => ({
  visitor: {
    JSXAttribute: {
      enter(path) {
        if (path.node.name === attribute) {
          path.remove()
        }
      },
    },
  },
})

export default stripAttribute
