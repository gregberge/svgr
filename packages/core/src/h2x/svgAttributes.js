import { JSXAttribute } from 'h2x-plugin-jsx'

const compareAttrsName = attr => otherAttr => attr.name === otherAttr.name
const compareAttrsValue = attr => otherAttr => attr.value === otherAttr.value

const compareAttrs = attr => otherAttr =>
  compareAttrsName(attr)(otherAttr) && compareAttrsValue(attr)(otherAttr)

const areAttrsAlreadyInjected = (node, attributes = {}) => {
  const nodeAttrs = node.attributes

  return Object.keys(attributes).reduce((accumulation, key) => {
    if (nodeAttrs.some(compareAttrs({ name: key, value: attributes[key] })))
      return accumulation
    return false
  }, true)
}

const svgAttribute = (attributes = {}) => () => {
  const keys = Object.keys(attributes)

  return {
    visitor: {
      JSXElement: {
        enter(path) {
          if (path.node.name !== 'svg') return
          if (areAttrsAlreadyInjected(path.node, attributes)) return

          const parseAttributes = keys.reduce((accumulation, key) => {
            const props = new JSXAttribute()
            props.name = key
            props.value = attributes[key]
            return [...accumulation, props]
          }, [])

          const mergeAttributes = path.node.attributes.reduce(
            (accumulation, value) => {
              if (accumulation.some(compareAttrsName(value)))
                return accumulation
              return [...accumulation, value]
            },
            parseAttributes,
          )

          path.node.attributes = mergeAttributes
          path.replace(path.node)
        },
      },
    },
  }
}

export default svgAttribute
