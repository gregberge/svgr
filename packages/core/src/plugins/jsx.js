import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
