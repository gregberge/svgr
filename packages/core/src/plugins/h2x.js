import { transform } from 'h2x-core'
import jsx from 'h2x-plugin-jsx'
import {
  emSize,
  expandProps,
  removeComments,
  removeDimensions,
  removeStyle,
  replaceAttrValues,
  stripAttribute,
  svgAttributes,
  svgRef,
  titleProp,
  toReactNative,
} from '..'

function configToPlugins(config) {
  const plugins = [
    jsx,
    stripAttribute('xmlns'),
    removeComments(),
    removeStyle(),
  ]
  if (config.replaceAttrValues)
    plugins.push(replaceAttrValues(config.replaceAttrValues))
  if (!config.dimensions) plugins.push(removeDimensions())
  if (config.icon) plugins.push(emSize())
  if (config.ref) plugins.push(svgRef())
  if (config.svgAttributes) plugins.push(svgAttributes(config.svgAttributes))
  if (config.expandProps) plugins.push(expandProps())
  if (config.native) plugins.push(toReactNative())
  if (config.titleProp) plugins.push(titleProp())
  return plugins
}

export default (code, config = {}, state = {}) => {
  const plugins = configToPlugins(config)
  return transform(code, {
    plugins,
    state,
    ...config.h2xConfig,
  })
}
