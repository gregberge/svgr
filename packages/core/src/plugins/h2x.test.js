import jsx from 'h2x-plugin-jsx'
import h2x from './h2x'
import stripAttribute from '../h2x/stripAttribute'

describe('h2x', () => {
  it('should take svg and convert it to jsx', () => {
    const result = h2x(
      `
      <svg>
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
    )

    expect(result).toMatchSnapshot()
  })

  it('should support config.h2xConfig', () => {
    const result = h2x(
      `
      <svg>
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
      { h2xConfig: { plugins: [jsx, stripAttribute('id')] } },
    )

    expect(result).toMatchSnapshot()
  })
})
