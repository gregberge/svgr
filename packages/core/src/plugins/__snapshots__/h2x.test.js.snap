// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`h2x should support config.h2xConfig 1`] = `
"<svg>
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" />
  </g>
</svg>
"
`;

exports[`h2x should take svg and convert it to jsx 1`] = `
"<svg>
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;
